### ISSUE TITLE
Unable to reproduce `Table 3 in the paper`

### ISSUE THREAD

**Issue by lerogo** [Liuziyu77/Visual-RFT#117]
Just like mentioned in the [`issue`](https://github.com/Liuziyu77/Visual-RFT/issues/97), we have made the necessary modifications to the files/scripts.

training log:  [training_log_table3.log](https://github.com/user-attachments/files/19320166/training_log_table3.log)

**The final output of the model is: `[]`**

**evaluation log** : [output_8cate.log](https://github.com/user-attachments/files/19320201/output_8cate.log)

training script:
```bash
timestamp=$1
echo "timestamp: ${timestamp}"

DATA_PATH=laolao77/ViRFT_COCO_8_cate_4_shot
CKPT_PATH=Qwen2-VL-2B-Instruct
SAVE_PATH="./output/Qwen2-VL-2B-Instruct_GRPO_COCO_8_cate_4_shot_${timestamp}"

export DEBUG_MODE="true" # Enable Debug if you want to see the rollout of model during RL
export LOG_PATH="./${SAVE_PATH}/debug_log.txt"

mkdir -p ${SAVE_PATH}

torchrun --master_addr ${MASTER_ADDR} --master-port ${MASTER_PORT} \
    --nnodes ${WORLD_SIZE} --node_rank ${RANK} --nproc-per-node=${GPUS} \
    src/virft/src/open_r1/grpo_classification.py \
    --output_dir ${SAVE_PATH}  \
    --model_name_or_path ${CKPT_PATH} \
    --dataset_name ${DATA_PATH} \
    --deepspeed src/virft/local_scripts/zero3.json \
    --max_prompt_length 1024 \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 2 \
    --logging_steps 1 \
    --bf16 \
    --report_to tensorboard \
    --gradient_checkpointing false \
    --attn_implementation flash_attention_2 \
    --max_pixels 401408 \
    --num_train_epochs **104** \
    --run_name "Qwen2-VL-2B_GRPO_COCO_8_cate_4shot_${timestamp}" \
    --save_steps 100 \
    --save_only_model true \
    --num_generations 8 \
    2>&1 | tee -a "./${SAVE_PATH}/training_log.log"
```

<img width="1318" alt="Image" src="https://github.com/user-attachments/assets/567f6e4c-4d73-4b60-97de-53c98f5e5c06" />

**Comment by Liuziyu77** [comment]
The issue you're describing—could it be due to using the wrong Python file? I see you're using the COCO dataset, but the reward function doesn't include IoU-reward, only Accuracy reward. `grpo.py` is for detection, while `grpo_classification.py` is for classification.

**Comment by lerogo** [comment]
> The issue you're describing—could it be due to using the wrong Python file? I see you're using the COCO dataset, but the reward function doesn't include IoU-reward, only Accuracy reward. `grpo.py` is for detection, while `grpo_classification.py` is for classification.

Thanks!
I am sorry, I will rerun this script tomorrow.

**Comment by lerogo** [comment]
We conducted the experiment again, and the results are as follows, which can replicate the results of Table 3 in the paper.

```bash
Key: bus, Value: 29.798253757223282
Key: train, Value: 41.41894994465757
Key: fire hydrant, Value: 43.43042357421661
Key: stop sign, Value: 35.9668673078038
Key: cat, Value: 59.94175072763568
Key: dog, Value: 47.27338345499707
Key: bed, Value: 25.881311670583706
Key: toilet, Value: 42.03987251015306
>>> print("mAP for selected categories: ", (AP_sum)/(len(selected_cate)))
mAP for selected categories:  40.71885161840885
```

**Comment by facetohard** [comment]
@lerogo Could you share your finial evaluation logs?I found the inference procedure takes too much time.

**Comment by lerogo** [comment]
> [@lerogo](https://github.com/lerogo) Could you share your finial evaluation logs?I found the inference procedure takes too much time.

generated_ids = model.generate(**inputs, max_new_tokens=1024, use_cache=True)

**Comment by facetohard** [comment]
@lerogo Thanks very much! The inference speed issue has been resolved. But I wonder that how many times do you run can reproduce Tabel 3. I got the exp res, which is 2 point percent lower than Tabel 3.

> > [@lerogo](https://github.com/lerogo) Could you share your finial evaluation logs?I found the inference procedure takes too much time.
> 
> generated_ids = model.generate(**inputs, max_new_tokens=1024, use_cache=True)

**Comment by YiqunChen1999** [comment]
> [@lerogo](https://github.com/lerogo) Thanks very much! The inference speed issue has been resolved. And another question that how many times do you run can reproduce Tabel 3. I got the exp res, which is 2 point percent lower than Tabel 3.
> 
> Key: bus, Value: 27.987041771906853
> Key: train, Value: 40.99657464155072
> Key: fire hydrant, Value: 41.521785179337094
> Key: stop sign, Value: 28.56397588568361
> Key: cat, Value: 61.65222685353163
> Key: dog, Value: 44.703722745317116
> Key: bed, Value: 23.834023465013317
> Key: toilet, Value: 40.51532839071756
> mAP for selected categories:  38.72183486663224
> > > [@lerogo](https://github.com/lerogo) Could you share your finial evaluation logs?I found the inference procedure takes too much time.
> > 
> > 
> > generated_ids = model.generate(**inputs, max_new_tokens=1024, use_cache=True)

@facetohard  Have you figured it out? I got 4 AP lower results:
```
Key: bus, Value: 25.853946201119776
Key: train, Value: 32.35729400326387
Key: fire hydrant, Value: 36.83383671827814
Key: stop sign, Value: 42.56665580279744
Key: cat, Value: 54.20205353475184
Key: dog, Value: 44.76382483900465
Key: bed, Value: 23.689937828985652
Key: toilet, Value: 30.252022853705796
mAP for selected categories:  36.3149464727384
```

**Comment by facetohard** [comment]
@YiqunChen1999 Maybe you can re-run same config a few times. I ran the experiment twice, and got mAP greater than table 3.
``` bash
Key: bus, Value: 30.631055284422683
Key: train, Value: 42.39880142616481
Key: fire hydrant, Value: 43.9506255832743
Key: stop sign, Value: 35.78792363994015
Key: cat, Value: 59.83318980579385
Key: dog, Value: 49.097848867812964
Key: bed, Value: 26.361232463812144
Key: toilet, Value: 43.01026536114605
mAP for selected categories:  41.38386780404586
```
and Another
```bash
Key: bus, Value: 34.08109958277205
Key: train, Value: 47.52282098514239
Key: fire hydrant, Value: 49.37425422313996
Key: stop sign, Value: 45.156891484478784
Key: cat, Value: 66.27715586200861
Key: dog, Value: 57.60788687230895
Key: bed, Value: 30.929416005762338
Key: toilet, Value: 43.97703627851085
mAP for selected categories:  46.86582016176549
```

**Comment by YiqunChen1999** [comment]
@facetohard Thank you for your advice. I think the RL/few-shot is sensitive to the random seed and it's important to provide the mean, std of results and the corresponding random seeds.

**Comment by BailiangJ** [comment]
Hi @lerogo @facetohard @YiqunChen1999 

Did you also reproduce the baseline numbers in Table 3?

![Image](https://github.com/user-attachments/assets/f5e4a854-f923-4735-a9ea-9630591ec4df)

I ran the inference and evaluation (with `use_cache=True` code directly on Qwen2-VL-2B-Instruct but I got

```
Key: bus, Value: 0.0
Key: train, Value: 16.46813161629079
Key: fire hydrant, Value: 0.0
Key: stop sign, Value: 0.0
Key: cat, Value: 30.182358230808966
Key: dog, Value: 0.0
Key: bed, Value: 0.0
Key: toilet, Value: 9.32806499180709
mAP for selected categories:  6.997319354863356
```

Thanks a lot in advance.

@Liuziyu77 Thanks a lot for sharing the code, am I using the correct baseline model?

**Comment by YiqunChen1999** [comment]
> Hi [@lerogo](https://github.com/lerogo) [@facetohard](https://github.com/facetohard) [@YiqunChen1999](https://github.com/YiqunChen1999)
> 
> Did you also reproduce the baseline numbers in Table 3?
> 
> ![Image](https://github.com/user-attachments/assets/f5e4a854-f923-4735-a9ea-9630591ec4df)
> 
> I ran the inference and evaluation (with `use_cache=True` code directly on Qwen2-VL-2B-Instruct but I got
> 
> ```
> Key: bus, Value: 0.0
> Key: train, Value: 16.46813161629079
> Key: fire hydrant, Value: 0.0
> Key: stop sign, Value: 0.0
> Key: cat, Value: 30.182358230808966
> Key: dog, Value: 0.0
> Key: bed, Value: 0.0
> Key: toilet, Value: 9.32806499180709
> mAP for selected categories:  6.997319354863356
> ```
> 
> Thanks a lot in advance.
> 
> [@Liuziyu77](https://github.com/Liuziyu77) Thanks a lot for sharing the code, am I using the correct baseline model?

Seems like the selected categories during inference are wrong (https://github.com/Liuziyu77/Visual-RFT/issues/149#issuecomment-2763153335).
